chain/network/src/store/schema/mod.rs

```diff
@@ -0,0 +1,279 @@
+pub struct KnownPeerStateRepr {
+    peer_info: primitives::PeerInfo,
+    status: KnownPeerStatus,
+    /// UNIX timestamps in nanos.
+    first_seen: u64,
+    last_seen: u64,
+}
+
+impl BorshRepr for KnownPeerStateRepr {
+    type T = primitives::KnownPeerState;
+    fn to_repr(s: &primitives::KnownPeerState) -> Self {
+        Self {
+            peer_info: s.peer_info.clone(),
+            status: s.status.clone().into(),
+            first_seen: s.first_seen,
+            last_seen: s.last_seen,
+        }
+    }
+
+    fn from_repr(s: Self) -> Result<primitives::KnownPeerState, Error> {
+        Ok(primitives::KnownPeerState {
+            peer_info: s.peer_info,
+            status: s.status.into(),
+            first_seen: s.first_seen,
+            last_seen: s.last_seen,
+        })
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct EdgeRepr {
+    key: (PeerId, PeerId),
+    nonce: u64,
+    signature0: Signature,
+    signature1: Signature,
+    removal_info: Option<(bool, Signature)>,
+}
+
+impl BorshRepr for EdgeRepr {
+    type T = primitives::Edge;
+
+    fn to_repr(e: &Self::T) -> Self {
+        Self {
+            key: e.key().clone(),
+            nonce: e.nonce(),
+            signature0: e.signature0().clone(),
+            signature1: e.signature1().clone(),
+            removal_info: e.removal_info().cloned(),
+        }
+    }
+    fn from_repr(e: Self) -> Result<Self::T, Error> {
+        Ok(primitives::Edge::new(e.key.0, e.key.1, e.nonce, e.signature0, e.signature1)
+            .with_removal_info(e.removal_info))
+    }
+}
+
+/////////////////////////////////////////////
+// Columns
+
+pub struct AccountAnnouncements;
+impl Column for AccountAnnouncements {
+    const COL: DBCol = DBCol::AccountAnnouncements;
+    type Key = AccountIdFormat;
+    type Value = Borsh<AnnounceAccount>;
+}
+
+pub struct Peers;
+impl Column for Peers {
+    const COL: DBCol = DBCol::Peers;
+    type Key = Borsh<PeerId>;
+    type Value = KnownPeerStateRepr;
+}
+
+pub struct PeerComponent;
+impl Column for PeerComponent {
+    const COL: DBCol = DBCol::PeerComponent;
+    type Key = Borsh<PeerId>;
+    type Value = Borsh<u64>;
+}
+
+pub struct ComponentEdges;
+impl Column for ComponentEdges {
+    const COL: DBCol = DBCol::ComponentEdges;
+    type Key = U64LE;
+    type Value = Vec<EdgeRepr>;
+}
+
+pub struct LastComponentNonce;
+impl Column for LastComponentNonce {
+    const COL: DBCol = DBCol::LastComponentNonce;
+    type Key = Borsh<()>;
+    type Value = Borsh<u64>;
+}
+
+////////////////////////////////////////////////////
+// Storage
+
+// Parsing error.
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Utf8(std::str::Utf8Error),
+    #[error(transparent)]
+    IO(std::io::Error),
+    #[error(transparent)]
+    Slice(std::array::TryFromSliceError),
+    #[error(transparent)]
+    AccountId(ParseAccountError),
+    #[error(transparent)]
+    Time(time::error::ComponentRange),
+}
```

Comment:
Mixed feeling here. This error is a union-of-types. When managing errors, we usually care not about types, but rather logical error conditions. So errors are more useful when they are a union of error conditions. In this like, the error would be 

```rust
enum Error {
  InvalidData { /* all the reasons why exactly bytes in store are not good */ },
  Io(io::Error), // sidenote: its Io, not IO, https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case
}
```

If cast in this light, it becomes obvious that

1. the comment is wrong, it's not necessary parsing error, genuine io error is also possible.
2. the `.commit` method can't produce `InvalidData`, so it perhaps can return just `io::Error`.

Another consideration is whether we need to actually treat invalid data errors and genuine errors differently. It seems that in either case this boils down to the underlying DB corruption? So, from application's point of view, this is just a single error condition? 

---

Taking the above considerations in mind, I'd probably keep it simple, and just use `io::Error` as the error type for this module, using `ErrorKind::InvalidData` to pack decoding errors: 

https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.InvalidData

Note that `io::Error::new(io::ErrorKind::InvalidData, concrete_error)` still preserves specific typed error inside. 

When thinking in terms of higher-level layers, we'd have two sources of IO errors -- from the store (which mean something is seriously bad and we should shut down) and from the network (which is just a normal course of operation). So perhaps the `scheme::Error` should be `io::Error`, and the `store::Error` should be 

```rust
/// Opaque error type representing storage errors.
///
/// Invariant: any store error is a critical operational operational error
/// which signals about data corruption. It wouldn't be wrong to replace all places /// where the error originates with outright panics. 
/// 
/// If you have an error condition which needs to be handles somehow, it should be 
/// some *other* error type. 
pub struct Error {
    inner: io::Error
}
```

Reply:
done